sentinel-adapter/sentinel-apache-dubbo-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/dubbo/SentinelDubboProviderFilter.java

```diff
@@ -85,18 +105,62 @@ public Result invoke(Invoker<?> invoker, Invocation invocation) throws RpcExcept
             return result;
         } catch (BlockException e) {
             return DubboAdapterGlobalConfig.getProviderFallback().handle(invoker, invocation, e);
-        } catch (RpcException e) {
-            Tracer.traceEntry(e, interfaceEntry);
-            Tracer.traceEntry(e, methodEntry);
-            throw e;
         } finally {
             if (methodEntry != null) {
                 methodEntry.exit(1, invocation.getArguments());
             }
             if (interfaceEntry != null) {
                 interfaceEntry.exit();
             }
-            ContextUtil.exit();
+        }
+    }
+
+
+    private Result asyncInvoke(Invoker<?> invoker, Invocation invocation, String interfaceResourceName, String methodResourceName) {
+        LinkedList<EntryHolder> queue = new LinkedList<>();
+        try {
+            queue.push(new EntryHolder(
+                    SphU.asyncEntry(interfaceResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.IN), null));
+            queue.push(new EntryHolder(
+                    SphU.asyncEntry(methodResourceName, ResourceTypeConstants.COMMON_RPC,
+                            EntryType.IN, invocation.getArguments()), invocation.getArguments()));
+            Result result = invoker.invoke(invocation);
+            result.whenCompleteWithContext((r, throwable) -> {
+                Throwable error = throwable;
+                if (error == null) {
+                    error = Optional.ofNullable(r).map(Result::getException).orElse(null);
+                }
+                while (!queue.isEmpty()) {
+                    EntryHolder holder = queue.pop();
+                    Tracer.traceEntry(error, holder.entry);
+                    exitEntry(holder);
+                }
+            });
+            return result;
+        } catch (BlockException e) {
+            while (!queue.isEmpty()) {
+                exitEntry(queue.pop());
+            }
+            return DubboAdapterGlobalConfig.getProviderFallback().handle(invoker, invocation, e);
+        }
```

Comment:
about the resource clean up(`exitEntry(holder)`),  had better move to a `finally` block which can guarantee it be called